### PR TITLE
Make `EnforceSorting` global sort aware, fix sort mis-optimizations involving unions, support parallel sort + merge transformations

### PR DESCRIPTION
**Pull request by mustafasrepo**
# Which issue does this PR close?

Closes [#5100](https://github.com/apache/arrow-datafusion/issues/5100).

# Rationale for this change

We want to have the ability to leverage parallelism in sorts when appropriate, which requires `EnforceSorting` to be partition aware. We also want to further decouple `EnforceSorting` and `EnforceDistribution` rules, with the end goal being complete orthogonality. Currently, there is coupling in both directions: In order to produce valid plans, one needs to apply the `EnforceDistribution` rule before applying `EnforceSorting`, and the `EnforceDistribution` breaks sorting requirements and produces invalid plans on its own.

With this PR, coupling is broken in one direction (`EnforceSorting` does not require `EnforceDistribution` to work correctly anymore), but not the other way yet (`EnforceDistribution` may still invalidate sort correctness, so one still needs to call `EnforceSorting` after that). Note that it is on our TODO list to fix the other direction too.

# What changes are included in this PR?

This PR adds a couple of functionalities to the `EnforceSorting` rule. 

- Adds parallelize option to the`EnforceSorting` rule. If this flag is enabled. Physical plans of the form
    
    ```rust
    "SortExec: [t1_id@0 ASC NULLS LAST]",
    "  CoalescePartitionsExec",
    "    RepartitionExec: partitioning=RoundRobinBatch(2), input_partitions=1",
    ```
    
    will turn into the plan below:
    
    ```rust
    "SortPreservingMergeExec: [t1_id@0 ASC NULLS LAST]",
    "  SortExec: [t1_id@0 ASC NULLS LAST]",
    "    RepartitionExec: partitioning=RoundRobinBatch(2), input_partitions=1",
    ```
    
    This increases speed in multi-threaded environments. You can see time comparisons below:
    
    | n_row | distinct | batch_size | Repartition Mode(8) | Rule On (mean) | Rule On(Median) | Rule Off (mean) | Rule Off(Median) |
    | --- | --- | --- | --- | --- | --- | --- | --- |
    | 1000 | 100 | 100 | RoundRobin | 3.015032ms | 2.988125ms | 5.213412ms | 4.515166ms |
    | 1000 | 100 | 100 | Hash | 6.00907ms | 5.895166ms | 10.636287ms | 10.3025ms |
    | 10_000 | 100 | 100 | RoundRobin | 24.429487ms | 24.042291ms | 84.282599ms | 84.117083ms |
    | 10_000 | 100 | 100 | Hash | 48.779278ms | 48.970708ms | 517.195637ms | 520.449625ms |
    | 20_000 | 100 | 100 | RoundRobin | 47.297341ms | 47.141041ms | 289.02502ms | 290.58625ms |
    | 20_000 | 100 | 100 | Hash | 114.230287ms | 107.842291ms | 1.992879062s | 2.002725333s |
    | 50_000 | 100 | 100 | RoundRobin | 117.185937ms | 115.300583ms | 1.604551912s | 1.609479958s |
    | 50_000 | 100 | 100 | Hash | 425.755595ms | 430.346625ms | 12.683566324s | 12.502681458s |
- Issue [#5100](https://github.com/apache/arrow-datafusion/issues/5100) is closed, fixing the limit bug.
- The `EnforceSorting` rule now also considers `SortPreservingMergeExec`s when performing optimizations. This enables us to optimize more queries.
- We handle cases where a `UnionExec` partially maintains the ordering of some of its children, such as the case below:
    
    ```rust
    "  UnionExec",
    "    SortExec: [nullable_col@0 ASC,non_nullable_col@1 ASC]",
    "      ParquetExec: limit=None, partitions={1 group: [[x]]}, projection=[nullable_col, non_nullable_col]",
    "    ParquetExec: limit=None, partitions={1 group: [[x]]}, output_ordering=[nullable_col@0 ASC], projection=[nullable_col, non_nullable_col]",
    ```
    
    In this case, output ordering of `UnionExec` is `nullable_col@0 ASC`. It maintains the ordering of its second child and partially maintains the ordering of its first child. The `EnforceSorting` rule now considers such partially-order-maintaining paths during optimization. 
    
- In order to handle multi-children operators like `UnionExec`s properly, the assumption that ordering only comes from a single path in the physical plan tree is removed. Now ordering can come from multiple paths.

# Are these changes tested?

We have added multiple tests to cover above mentioned functionalities. Approximately 550 lines of the changes come from test or test utils.

# Are there any user-facing changes?

No.

### CODE REVIEW

**Comment by ozankabak** [comment]
After sending updates to the base branch, this went out of sync. I resolved the merge conflicts to accelerate you, but please verify/double check before building on top of this.

**Comment by ozankabak** [comment]
I did an initial round of review on this, and subsequently made some simplifications and added comments. I have a few questions we should go over in our next meeting.

As I was changing the base branch (since the dependent API change branch merged upstream already), I unfortunately lost our commit history, so we will be starting from scratch (in terms of history) but that's OK.

**Comment by mingmwang** [comment]
I plan to do some refactoring work to the `EnforceSorting` rule. I think currently it was too complex to understand.

**Comment by ozankabak** [comment]
> I plan to do some refactoring work to the `EnforceSorting` rule. I think currently it was too complex to understand.

@mingmwang: Sounds good, as long as we don't lose functionality I am always on board with refactors. This PR has some important additions, let us finish this (a couple days at most) and add some more tests to avoid losing any functionality during refactors. With that, you can safely take a look at how you can simplify it and refactor.

**Comment by mingmwang** [comment]
I can understand the final purpose of this PR. I will review this PR carefully tomorrow.
One thing that I would like to hear from you guys is for the `SortExec`,  should we add a flag to indicate it is local sort or global sort ?  This will make the final physical plans easy to  understand. And when we do Sort related optimizations, it is less error prone.

@mustafasrepo @ozankabak @alam

**Comment by mustafasrepo** [comment]
> I can understand the final purpose of this PR. I will review this PR carefully tomorrow. One thing that I would like to hear from you guys is for the `SortExec`, should we add a flag to indicate it is local sort or global sort ? This will make the final physical plans easy to understand. And when we do Sort related optimizations, it is less error prone.
> 
> @mustafasrepo @ozankabak

Currently we check if `SortExec` preserves partitioning and its input has partitioning size 1; to decide for whether SortExec is global or local. I think using a flag would be useful.

File: datafusion/core/src/physical_optimizer/sort_enforcement.rs [hunk 1]
```diff
@@ -897,6 +1258,236 @@ mod tests {
...
+            "  UnionExec",
+            "    SortExec: [nullable_col@0 ASC]",
+            "      ParquetExec: limit=None, partitions={1 group: [[x]]}, projection=[nullable_col, non_nullable_col]",
+            "    SortExec: [nullable_col@0 ASC]",
+            "      ParquetExec: limit=None, partitions={1 group: [[x]]}, projection=[nullable_col, non_nullable_col]",
+        ];
+        assert_optimized!(expected_input, expected_optimized, physical_plan);
+        Ok(())
+    }
+
```
Comment (mingmwang) [inline datafusion/core/src/physical_optimizer/sort_enforcement.rs:1397]: I think the original UnionExec's output ordering can satisfy the SortPreservingMergeExec's sort requirements already, why need to change the plan ?

File: datafusion/core/src/physical_optimizer/sort_enforcement.rs [hunk 2]
```diff
@@ -151,25 +325,95 @@ impl PhysicalOptimizerRule for EnforceSorting {
...
+/// This function turns plans of the form
+///      "SortExec: [a@0 ASC]",
+///      "  CoalescePartitionsExec",
+///      "    RepartitionExec: partitioning=RoundRobinBatch(8), input_partitions=1",
+/// to
+///      "SortPreservingMergeExec: [a@0 ASC]",
+///      "  SortExec: [a@0 ASC]",
+///      "    RepartitionExec: partitioning=RoundRobinBatch(8), input_partitions=1",
+/// by following connections from `CoalescePartitionsExec`s to `SortExec`s.
+/// By performing sorting in parallel, we can increase performance in some scenarios.
```
Comment (mingmwang) [inline datafusion/core/src/physical_optimizer/sort_enforcement.rs:337]: There is already a rule `GlobalSortSelection` which do the similar staff,  I think if the rule is not owner enough, you guys can add the sort selection logic to that rule.

**Comment by mingmwang** [comment]
In your branch, could you please verify those cases?

```rust
async fn test_remove_unnecessary_sort5() -> Result<()> {
        let schema = create_test_schema()?;
        let source = memory_exec(&schema);
        let sort_exprs = vec![sort_expr("non_nullable_col", &schema)];
        let sort = sort_exec(sort_exprs.clone(), source);
        let spm = sort_preserving_merge_exec(sort_exprs, sort);

        let sort_exprs = vec![
            sort_expr("nullable_col", &schema),
            sort_expr("non_nullable_col", &schema),
        ];
        let sort2 = sort_exec(sort_exprs.clone(), spm);
        let spm2 = sort_preserving_merge_exec(sort_exprs, sort2);

        let sort_exprs = vec![sort_expr("nullable_col", &schema)];
        let sort3 = sort_exec(sort_exprs.clone(), spm2);
        let physical_plan = repartition_exec(repartition_exec(sort3));

        let expected_input = vec![
            "RepartitionExec: partitioning=RoundRobinBatch(10), input_partitions=10",
            "  RepartitionExec: partitioning=RoundRobinBatch(10), input_partitions=1",
            "    SortExec: [nullable_col@0 ASC], global=true",
            "      SortPreservingMergeExec: [nullable_col@0 ASC,non_nullable_col@1 ASC]",
            "        SortExec: [nullable_col@0 ASC,non_nullable_col@1 ASC], global=true",
            "          SortPreservingMergeExec: [non_nullable_col@1 ASC]",
            "            SortExec: [non_nullable_col@1 ASC], global=true",
            "              MemoryExec: partitions=0, partition_sizes=[]",
        ];

        let expected_optimized = vec![
            "RepartitionExec: partitioning=RoundRobinBatch(10), input_partitions=10",
            "  RepartitionExec: partitioning=RoundRobinBatch(10), input_partitions=0",
            "    MemoryExec: partitions=0, partition_sizes=[]",
        ];
        assert_optimized!(expected_input, expected_optimized, physical_plan);
        Ok(())
    }

```

```rust
#[tokio::test]
    async fn test_do_not_remove_sort_with_limit() -> Result<()> {
        let schema = create_test_schema()?;

        let source1 = parquet_exec(&schema);
        let sort_exprs = vec![
            sort_expr("nullable_col", &schema),
            sort_expr("non_nullable_col", &schema),
        ];
        let sort = sort_exec(sort_exprs.clone(), source1);
        let limit = limit_exec(sort);

        let parquet_sort_exprs = vec![sort_expr("nullable_col", &schema)];
        let source2 = parquet_exec_sorted(&schema, parquet_sort_exprs);

        let union = union_exec(vec![source2, limit]);
        let repartition = repartition_exec(union);
        let physical_plan = sort_preserving_merge_exec(sort_exprs, repartition);

        let expected_input = vec![
            "SortPreservingMergeExec: [nullable_col@0 ASC,non_nullable_col@1 ASC]",
            "  RepartitionExec: partitioning=RoundRobinBatch(10), input_partitions=2",
            "    UnionExec",
            "      ParquetExec: limit=None, partitions={1 group: [[x]]}, output_ordering=[nullable_col@0 ASC], projection=[nullable_col, non_nullable_col]",
            "      GlobalLimitExec: skip=0, fetch=100",
            "        LocalLimitExec: fetch=100",
            "          SortExec: [nullable_col@0 ASC,non_nullable_col@1 ASC], global=true",
            "            ParquetExec: limit=None, partitions={1 group: [[x]]}, projection=[nullable_col, non_nullable_col]",
        ];

        // expect to keep the bottom SortExec
        let expected_optimized = vec![
            "SortPreservingMergeExec: [nullable_col@0 ASC,non_nullable_col@1 ASC]",
            "  SortExec: [nullable_col@0 ASC,non_nullable_col@1 ASC], global=false",
            "    RepartitionExec: partitioning=RoundRobinBatch(10), input_partitions=2",
            "      UnionExec",
            "        ParquetExec: limit=None, partitions={1 group: [[x]]}, output_ordering=[nullable_col@0 ASC], projection=[nullable_col, non_nullable_col]",
            "        GlobalLimitExec: skip=0, fetch=100",
            "          LocalLimitExec: fetch=100",
            "            SortExec: [nullable_col@0 ASC,non_nullable_col@1 ASC], global=true",
            "              ParquetExec: limit=None, partitions={1 group: [[x]]}, projection=[nullable_col, non_nullable_col]",
        ];
        assert_optimized!(expected_input, expected_optimized, physical_plan);
        Ok(())
    }
```

File: datafusion/core/src/physical_optimizer/sort_enforcement.rs (hunk 1 above)
Comment (mustafasrepo) [reply datafusion/core/src/physical_optimizer/sort_enforcement.rs]: One of the `SortExec`s under the `UnionExec` is unnecesarily finer than the required Sorting by above executor. Hence unnecessarily fine SortExec is changed with the absolutely necessary `SortExec` by above operators.

**Comment by mustafasrepo** [comment]
> In your branch, could you please verify those cases?
> 
> ```rust
> async fn test_remove_unnecessary_sort5() -> Result<()> {
>         let schema = create_test_schema()?;
>         let source = memory_exec(&schema);
>         let sort_exprs = vec![sort_expr("non_nullable_col", &schema)];
>         let sort = sort_exec(sort_exprs.clone(), source);
>         let spm = sort_preserving_merge_exec(sort_exprs, sort);
> 
>         let sort_exprs = vec![
>             sort_expr("nullable_col", &schema),
>             sort_expr("non_nullable_col", &schema),
>         ];
>         let sort2 = sort_exec(sort_exprs.clone(), spm);
>         let spm2 = sort_preserving_merge_exec(sort_exprs, sort2);
> 
>         let sort_exprs = vec![sort_expr("nullable_col", &schema)];
>         let sort3 = sort_exec(sort_exprs.clone(), spm2);
>         let physical_plan = repartition_exec(repartition_exec(sort3));
> 
>         let expected_input = vec![
>             "RepartitionExec: partitioning=RoundRobinBatch(10), input_partitions=10",
>             "  RepartitionExec: partitioning=RoundRobinBatch(10), input_partitions=1",
>             "    SortExec: [nullable_col@0 ASC], global=true",
>             "      SortPreservingMergeExec: [nullable_col@0 ASC,non_nullable_col@1 ASC]",
>             "        SortExec: [nullable_col@0 ASC,non_nullable_col@1 ASC], global=true",
>             "          SortPreservingMergeExec: [non_nullable_col@1 ASC]",
>             "            SortExec: [non_nullable_col@1 ASC], global=true",
>             "              MemoryExec: partitions=0, partition_sizes=[]",
>         ];
> 
>         let expected_optimized = vec![
>             "RepartitionExec: partitioning=RoundRobinBatch(10), input_partitions=10",
>             "  RepartitionExec: partitioning=RoundRobinBatch(10), input_partitions=0",
>             "    MemoryExec: partitions=0, partition_sizes=[]",
>         ];
>         assert_optimized!(expected_input, expected_optimized, physical_plan);
>         Ok(())
>     }
> ```
> 
> ```rust
> #[tokio::test]
>     async fn test_do_not_remove_sort_with_limit() -> Result<()> {
>         let schema = create_test_schema()?;
> 
>         let source1 = parquet_exec(&schema);
>         let sort_exprs = vec![
>             sort_expr("nullable_col", &schema),
>             sort_expr("non_nullable_col", &schema),
>         ];
>         let sort = sort_exec(sort_exprs.clone(), source1);
>         let limit = limit_exec(sort);
> 
>         let parquet_sort_exprs = vec![sort_expr("nullable_col", &schema)];
>         let source2 = parquet_exec_sorted(&schema, parquet_sort_exprs);
> 
>         let union = union_exec(vec![source2, limit]);
>         let repartition = repartition_exec(union);
>         let physical_plan = sort_preserving_merge_exec(sort_exprs, repartition);
> 
>         let expected_input = vec![
>             "SortPreservingMergeExec: [nullable_col@0 ASC,non_nullable_col@1 ASC]",
>             "  RepartitionExec: partitioning=RoundRobinBatch(10), input_partitions=2",
>             "    UnionExec",
>             "      ParquetExec: limit=None, partitions={1 group: [[x]]}, output_ordering=[nullable_col@0 ASC], projection=[nullable_col, non_nullable_col]",
>             "      GlobalLimitExec: skip=0, fetch=100",
>             "        LocalLimitExec: fetch=100",
>             "          SortExec: [nullable_col@0 ASC,non_nullable_col@1 ASC], global=true",
>             "            ParquetExec: limit=None, partitions={1 group: [[x]]}, projection=[nullable_col, non_nullable_col]",
>         ];
> 
>         // expect to keep the bottom SortExec
>         let expected_optimized = vec![
>             "SortPreservingMergeExec: [nullable_col@0 ASC,non_nullable_col@1 ASC]",
>             "  SortExec: [nullable_col@0 ASC,non_nullable_col@1 ASC], global=false",
>             "    RepartitionExec: partitioning=RoundRobinBatch(10), input_partitions=2",
>             "      UnionExec",
>             "        ParquetExec: limit=None, partitions={1 group: [[x]]}, output_ordering=[nullable_col@0 ASC], projection=[nullable_col, non_nullable_col]",
>             "        GlobalLimitExec: skip=0, fetch=100",
>             "          LocalLimitExec: fetch=100",
>             "            SortExec: [nullable_col@0 ASC,non_nullable_col@1 ASC], global=true",
>             "              ParquetExec: limit=None, partitions={1 group: [[x]]}, projection=[nullable_col, non_nullable_col]",
>         ];
>         assert_optimized!(expected_input, expected_optimized, physical_plan);
>         Ok(())
>     }
> ```

Thanks for the suggestion @mingmwang. I have tried the tests, second test already passes. However, first test is not producing optimized result. I will add this functionality and let you know when it is ready. Thanks for pointing it out.

**Comment by mingmwang** [comment]
> Thanks for the suggestion @mingmwang. I have tried the tests, second test already passes. However, first test is not producing optimized result. I will add this functionality and let you know when it is ready. Thanks for pointing it out.

I think you can leave it here. I had almost finished my refactoring work, it is based the on Top-Down approach and can support very sophisticated optimizations.

**Comment by mingmwang** [comment]
I think the current Sort removing logic is problematic. It is very dangerous to remove the top `SortExec` because it will
change the final result ordering.  For example, in the below case, the original final output should be ordered by the column
`non_nullable_col` in `ASC`, but after the optimization, the final result is ordered by that column in `DESC`.

I think a good optimization is that the final plan's output ordering should be compatible with the original plan's output ordering. Compatible means we allow a finer output ordering, but we should not change it to different things or remove them. The global `SortExec` is usually added by the users,  we should not change the behavior and break user's expectation.
 

```rust
let expected_input = vec![
            "WindowAggExec: wdw=[count: Ok(Field { name: \"count\", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Range, start_bound: Preceding(NULL), end_bound: CurrentRow }]",
            "  FilterExec: NOT non_nullable_col@1",
            "    SortExec: [non_nullable_col@1 ASC NULLS LAST]",
            "      WindowAggExec: wdw=[count: Ok(Field { name: \"count\", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Range, start_bound: Preceding(NULL), end_bound: CurrentRow }]",
            "        SortExec: [non_nullable_col@1 DESC]",
            "          MemoryExec: partitions=0, partition_sizes=[]",
        ];

        let expected_optimized = vec![
            "WindowAggExec: wdw=[count: Ok(Field { name: \"count\", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Range, start_bound: CurrentRow, end_bound: Following(NULL) }]",
            "  FilterExec: NOT non_nullable_col@1",
            "    WindowAggExec: wdw=[count: Ok(Field { name: \"count\", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Range, start_bound: Preceding(NULL), end_bound: CurrentRow }]",
            "      SortExec: [non_nullable_col@1 DESC]",
            "        MemoryExec: partitions=0, partition_sizes=[]",
        ];

```

**Comment by mustafasrepo** [comment]
> I think the current Sort removing logic is problematic. It is very dangerous to remove the top `SortExec` because it will change the final result ordering. For example, in the below case, the original final output should be ordered by the column `non_nullable_col` in `ASC`, but after the optimization, the final result is ordered by that column in `DESC`.
> 
> I think a good optimization is that the final plan's output ordering should be compatible with the original plan's output ordering. Compatible means we allow a finer output ordering, but we should not change it to different things or remove them. The global `SortExec` is usually added by the users, we should not change the behavior and break user's expectation.
> 
> ```rust
> let expected_input = vec![
>             "WindowAggExec: wdw=[count: Ok(Field { name: \"count\", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Range, start_bound: Preceding(NULL), end_bound: CurrentRow }]",
>             "  FilterExec: NOT non_nullable_col@1",
>             "    SortExec: [non_nullable_col@1 ASC NULLS LAST]",
>             "      WindowAggExec: wdw=[count: Ok(Field { name: \"count\", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Range, start_bound: Preceding(NULL), end_bound: CurrentRow }]",
>             "        SortExec: [non_nullable_col@1 DESC]",
>             "          MemoryExec: partitions=0, partition_sizes=[]",
>         ];
> 
>         let expected_optimized = vec![
>             "WindowAggExec: wdw=[count: Ok(Field { name: \"count\", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Range, start_bound: CurrentRow, end_bound: Following(NULL) }]",
>             "  FilterExec: NOT non_nullable_col@1",
>             "    WindowAggExec: wdw=[count: Ok(Field { name: \"count\", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Range, start_bound: Preceding(NULL), end_bound: CurrentRow }]",
>             "      SortExec: [non_nullable_col@1 DESC]",
>             "        MemoryExec: partitions=0, partition_sizes=[]",
>         ];
> ```

I agree that we shouldn't violate user expectations about query. Specifically when user enters `GlobalSortExec` output ordering should be aligned with it. The above test case is constructed synthetically. As far as I am aware current rule doesn't violate output ordering of `GlobalSortExec`. If you have any query that you think its optimized plan is wrong or misleading, please mention it. I will be happy to fix it. Thanks for your feedback @mingmwang

**Comment by ozankabak** [comment]
Hi @mingmwang, we will do a one final round of changes on this and we will send it upstream. Once this merges and gives all of us a reference point in terms of functionality, let's switch roles and collaborate on your refactor 🙂  We can see if the two approaches are equivalent in terms of functionality. If they are indeed equivalent, or your refactor is even more powerful, then we can get yours merged as a follow-on to this; this way we would have the complete track record of development in the commit history. Moreover, we can compare and contrast in the future too, do bisects, see if there is any bugs or something else we miss today.

Thanks a lot for providing test cases and reminders on important issues, we really appreciate it. I will discuss with @mustafasrepo in detail tomorrow about making sure that this never interferes with global sorting. As you suggest, adding a flag to sorts indicating global status may be safe route to take. In the meantime, if you have an example query where this issue may happen, please share with us. We plan to finalize this and get back to you tomorrow.

**Review by ozankabak — APPROVED** [review]
For posterity, this went through our review process and is now reviewed by upstream.

**Comment by ozankabak** [comment]
Closing as it merged upstream.

**Pull request closed without merging**